### PR TITLE
Limit dnd operations on desktop icons

### PR DESCRIPTION
**Pull request by robertxgray**
Some drag and drop operations involving desktop icon files are not defined and produce an error as result. IMHO the message errors are meaningless for most users and their experience would be better if such operations cannot be performed.

Expected behaviour with this change:

- Computer and network icons cannot be used as a drop target.
- Desktop icon files cannot be dropped on other files or folders.

### CODE REVIEW

**Review by rbuj — CHANGES_REQUESTED** [review]
no longer shows the message to the user about the operation is not supported.

Screenshot at 2021-03-25 13-57-33

**Comment by rbuj** [comment]
It involves other locations, including places side panel.

File: libcaja-private/caja-dnd.c [hunk 1]
```diff
@@ -457,12 +459,31 @@ caja_drag_default_drop_action_for_icons (GdkDragContext *context,
     dropped_file = caja_file_get_existing_by_uri (dropped_uri);
     target_file = caja_file_get_existing_by_uri (target_uri_string);
 
+    if (eel_uri_is_desktop (dropped_uri))
```
Comment (rbuj) [inline libcaja-private/caja-dnd.c:462]: If `eel_uri_is_desktop(uri)` returned TRUE means that the uri starts with `x-caja-desktop:` prefix, for instance the **Computer** icon placed on user's desktop (x-caja-desktop:///computer)

**Review by rbuj — CHANGES_REQUESTED** [review]
drag and drop modifier keys do not work:
- GDK_ACTION_LINK is enforced by Alt key
- GDK_ACTION_MOVE is enforced by Shift key
- GDK_ACTION_COPY is enforced by Ctrl key

Edit: The modifier keys work as expected, GDK_ACTION_COPY is enforced by pressing Ctrl+Shift.

File: libcaja-private/caja-dnd.c (hunk 1 above)
Comment (robertxgray) [reply libcaja-private/caja-dnd.c:462]: That code is related to the second expected behaviour.

- Desktop icon files cannot be dropped on other files or folders.

With desktop icon files I mean any CajaDesktopIconFile, as described in caja-desktop-icon-file.h. My intention was to stop the user from performing dnd operations that lead to a sure error and leave the dnd animation as the only feedback. I think that improves the user experience, but I don't know if it's a majority opinion.

**Comment by robertxgray** [comment]
To be honest I didn't know about those modifier keys, but I just performed some tests with both my proposal and the original code and couldn't find any problem. Can you provide more information about the use cases where the modifier keys are not working please?

File: libcaja-private/caja-dnd.c [hunk 2]
```diff
@@ -457,12 +459,31 @@ caja_drag_default_drop_action_for_icons (GdkDragContext *context,
...
+        caja_file_unref (target_file);
+        return;
+    }
     /*
      * Check for trash URI.  We do a find_directory for any Trash directory.
      * Passing 0 permissions as mate-vfs would override the permissions
      * passed with 700 while creating .Trash directory
      */
-    if (eel_uri_is_trash (target_uri_string))
+    else if (eel_uri_is_trash (target_uri_string))
```
Comment (rbuj) [inline libcaja-private/caja-dnd.c:486]: It still shows the error dialog when dragging an icon with the x-caja-desktop schema to the Trash.

**Comment by rbuj** [comment]
> To be honest I didn't know about those modifier keys, but I just performed some tests with both my proposal and the original code and couldn't find any problem. Can you provide more information about the use cases where the modifier keys are not working please?

Don't worry, when I tried to enforce the link creation I pressed the Alt key, however to force a link creation you have to press Shift and Control keys. I'm sorry.

File: libcaja-private/caja-dnd.c [hunk 3]
```diff
@@ -491,20 +512,6 @@ caja_drag_default_drop_action_for_icons (GdkDragContext *context,
...
-            if (actions & GDK_ACTION_MOVE)
-            {
-                *action = GDK_ACTION_MOVE;
-            }
-
-            g_object_unref (target);
-            caja_file_unref (dropped_file);
-            caja_file_unref (target_file);
-            return;
-        }
```
Comment (rbuj) [inline libcaja-private/caja-dnd.c:507]: This makes #1502 unnecessary.

File: libcaja-private/caja-dnd.c (hunk 1 above)
Comment (rbuj) [reply libcaja-private/caja-dnd.c:462]: IMHO it should be handled by gvfs backends. `caja_drag_default_drop_action_for_icons` sets the default dnd action, when it's available, and it doesn't stop/end a dnd job by setting `*action = 0;`

**Comment by robertxgray** [comment]
Okay, I'll try to implement the change in gvfs backends.

**Pull request closed without merging**